From 8ec7aa963ee6c53e91efc0b6e259d9eb0c91e1a1 Mon Sep 17 00:00:00 2001
From: Eric Sauvageau <merlin@asuswrt-merlin.net>
Date: Sun, 28 Jun 2026 23:42:44 -0400
Subject: [PATCH] rc: don't assign the whole /56 to br0 if that's what you get
 - limit to a /64

This reproduces the fix from the original prefix patch without affecting SDN.

(cherry picked from commit 298d52ec096989fbc4143798303daa91bdfeced5)
---
 release/src/router/rc/services.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/release/src/router/rc/services.c b/release/src/router/rc/services.c
index b5b219d816..e6b5ef8d26 100644
--- a/release/src/router/rc/services.c
+++ b/release/src/router/rc/services.c
@@ -2861,13 +2861,17 @@ void add_ip6_lanaddr(void)
 	char ip[INET6_ADDRSTRLEN + 4];
 #endif
 	const char *p;
+	int prefix_len;
 
 #ifdef RTCONFIG_MULTILAN_CFG
 	p = _add_ip6_lansdnaddr();
 #else
 	p = ipv6_router_address(NULL);
 	if (*p) {
-		snprintf(ip, sizeof(ip), "%s/%d", p, nvram_get_int(ipv6_nvname("ipv6_prefix_length")) ? : 64);
+		prefix_len = nvram_get_int(ipv6_nvname("ipv6_prefix_length"));
+		if (prefix_len <= 56)
+			prefix_len = 64;
+		snprintf(ip, sizeof(ip), "%s/%d", p, prefix_len);
 		eval("ip", "-6", "addr", "add", ip, "dev", nvram_safe_get("lan_ifname"));
 		if (!nvram_match(ipv6_nvname("ipv6_rtr_addr"), (char*)p))
 			nvram_set(ipv6_nvname("ipv6_rtr_addr"), (char*)p);
-- 
2.25.1

